File: python/MASTER_SLAVE_WHOLE_PART_ABSTRACT_FACTORY/MASTER_SLAVE_WHOLE_PART_ABSTRACT_FACTORY/abstract_factory_master/ReservacionMasterFactory.py
from .ProductoCineFactory import ProductoCineFactory

from ..whole_part.ReservacionSencilla       import ReservacionSencilla
from ..whole_part.ReservacionVIP            import ReservacionVIP
from ..master_slaves.GestorDatosCliente     import GestorDatosCliente
from ..master_slaves.GestorDatosCompra      import GestorDatosCompra
from ..master_slaves.GestorDatosFuncion     import GestorDatosFuncion
from ..master_slaves.GestorDatosBeneficios  import GestorDatosBeneficios

class ReservacionMasterFactory(ProductoCineFactory):

    def invocaSubordinados_sencillo(self, nombres, apellidoPaterno,
                     apellidoMaterno, edad, tipoCliente,
                     numeroTarjeta, fechaVencimiento,
                     asiento, horario, pelicula):
        datosCliente = GestorDatosCliente().preparaDatosCliente(nombres, apellidoPaterno, apellidoMaterno, edad, tipoCliente)
        datosCompra  = GestorDatosCompra().preparaDatosCompra(numeroTarjeta, fechaVencimiento)
        datosFuncion = GestorDatosFuncion().preparaDatosFuncion(asiento, horario, pelicula)
        return datosCliente,datosCompra,datosFuncion
    
    def invocaSubordinados_vip(self, nombres, apellidoPaterno,
                     apellidoMaterno, edad, tipoCliente,
                     numeroTarjeta, fechaVencimiento,
                     asiento, horario, pelicula,
                     beneficios):
        datosCliente    = GestorDatosCliente().preparaDatosCliente(nombres, apellidoPaterno, apellidoMaterno, edad, tipoCliente)
        datosCompra     = GestorDatosCompra().preparaDatosCompra(numeroTarjeta, fechaVencimiento)
        datosFuncion    = GestorDatosFuncion().preparaDatosFuncion(asiento, horario, pelicula)
        datosBeneficios = GestorDatosBeneficios().preparaDatosBeneficios(beneficios)
        return datosCliente,datosCompra,datosFuncion,datosBeneficios
    
    def crearProductoCineSencillo(self,nombres,
                       apellidoPaterno ,
                       apellidoMaterno ,
                       edad,
                       tipoCliente,
                       numeroTarjeta,
                       fechaVencimiento,
                       asiento,
                       horario,
                       pelicula):
        datosCliente,datosCompra,datosFuncion = self.invocaSubordinados_sencillo(nombres, apellidoPaterno,
                          apellidoMaterno, edad, tipoCliente,
                          numeroTarjeta, fechaVencimiento,
                          asiento, horario, pelicula)
        return ReservacionSencilla(datosCliente,datosCompra,datosFuncion)
    
    def crearProductoCineVIP(self,nombres,
                       apellidoPaterno ,
                       apellidoMaterno ,
                       edad,
                       tipoCliente,
                       numeroTarjeta,
                       fechaVencimiento,
                       asiento,
                       horario,
                       pelicula,
                       beneficios):
        datosCliente,datosCompra,datosFuncion,datosBeneficios= self.invocaSubordinados_vip(nombres, apellidoPaterno,
                          apellidoMaterno, edad, tipoCliente,
                          numeroTarjeta, fechaVencimiento,
                          asiento, horario, pelicula,
                          beneficios)
        return ReservacionVIP(datosCliente,datosCompra,datosFuncion,datosBeneficios)

File: php/MASTER_SLAVE_WHOLE_PART_ABSTRACT_FACTORY/principal/ReservacionMasterFactory.php
<?php
require './gestores/GestorDatosBeneficio.php';


/**
 * Description of ReservacionMasterFactory
 *
 * @author giltenorio
 */
class ReservacionMasterFactory implements ProductoCineFactory {

    public $reservacion;

    //put your code here
    public function crearProductoCineSencillo($nombre, $apellidoP, $apellidoM, $edad, $tipoC, $numTarjeta,
            $fecha, $asiento, $dia, $mes, $anio, $pelicula) {
        $this->invocaSubordinadosSencillos($nombre, $apellidoP, $apellidoM, $edad, $tipoC, $numTarjeta, $fecha, $asiento, $dia, $mes, $anio, $pelicula);
        return $this->reservacion;
    }

    public function crearProductoCineVIP($nombre, $apellidoP, $apellidoM, $edad, $tipoC, $numTarjeta,
            $fecha, $asiento, $dia, $mes, $anio, $pelicula, $beneficio) {
        
        $this->invocaSubordinadosVIP($nombre, $apellidoP, $apellidoM, $edad, $tipoC, $numTarjeta, $fecha, $asiento, $dia, $mes, $anio, $pelicula, $beneficio);
        
        return $this->reservacion;
    }

    public function invocaSubordinadosSencillos($nombre, $apellidoP, $apellidoM, $edad, $tipoC, $numTarjeta,
            $fecha, $asiento, $dia, $mes, $anio, $pelicula) {

        $datosCliente = GestorDatosCliente::preparaDatosCliente($nombre, $apellidoP, $apellidoM, $edad, $tipoC);
        $datosCompra = GestorDatosCompra::preparaDatosCompra($numTarjeta, $fecha);
        $datosFuncion = GestorDatosFuncion::preparaDatosFuncion($asiento, $dia, $mes, $anio, $pelicula);
        $this->reservacion = new ReservacionSencilla($datosCliente, $datosCompra, $datosFuncion);
    }
    
    public function invocaSubordinadosVIP($nombre, $apellidoP, $apellidoM, $edad, $tipoC, $numTarjeta,
            $fecha, $asiento, $dia, $mes, $anio, $pelicula, $beneficio) {
        
        $datosCliente = GestorDatosCliente::preparaDatosCliente($nombre, $apellidoP, $apellidoM, $edad, $tipoC);
        $datosCompra = GestorDatosCompra::preparaDatosCompra($numTarjeta, $fecha);
        $datosFuncion = GestorDatosFuncion::preparaDatosFuncion($asiento, $dia, $mes, $anio, $pelicula);
       
        $datosBeneficio = GestorDatosBeneficio::preparaDatosBeneficio($beneficio);
        
        $this->reservacion = new ReservacionVIP($datosCliente, $datosCompra, $datosFuncion, $datosBeneficio);
        
    }

}
